apps/functions-subscription/src/use-cases/insert-subscription.ts

```diff
@@ -0,0 +1,48 @@
+import { pipe } from 'fp-ts/lib/function';
+import * as O from 'fp-ts/Option';
+import * as TE from 'fp-ts/TaskEither';
+import * as RTE from 'fp-ts/ReaderTaskEither';
+import { Capabilities } from '../domain/capabilities';
+import { TrialId, UserId, makeSubscriptionId } from '../domain/subscription';
+import { SubscriptionAlreadyExists } from './errors';
+
+// Maps all the requirements for this use-case
+type Env = Pick<
+  Capabilities,
+  | 'subscriptionRequestWriter'
+  | 'subscriptionReader'
+  | 'subscriptionWriter'
+  | 'hashFn'
+>;
+
+export const insertSubscription = (userId: UserId, trialId: TrialId) =>
+  pipe(
+    RTE.ask<Env>(),
+    RTE.apSW('id', makeSubscriptionId(trialId, userId)),
+    RTE.flatMapTaskEither(
+      ({
+        subscriptionReader,
+        subscriptionWriter,
+        subscriptionRequestWriter,
+        id,
+      }) =>
+        pipe(
+          subscriptionReader.get(id),
+          TE.flatMap(
+            O.fold(
+              () =>
+                pipe(
+                  subscriptionRequestWriter.insert({ userId, trialId }),
+                  TE.flatMap(() =>
+                    subscriptionWriter.insert({ userId, trialId, id }),
+                  ),
+                ),
```

Comment:
what about extracting this third level pipe into its own method? ie.

```ts
const handleMissingSubscription = (userId: UserId, trialId: TrialId, id: SubscriptionId, env: Env) =>
  pipe(
    env.subscriptionRequestWriter.insert({ userId, trialId }),
    TE.flatMap(() => env.subscriptionWriter.insert({ userId, trialId, id }))
  );
```

Reply:
✅ Based on your suggestion, I've made some additional refinements to the method